cpp/src/io/csv/writer_impl.cu

```diff
@@ -0,0 +1,545 @@
+
+    // host algorithm call, where the underlying call
+    // is also host_write taking a host buffer;
+    //
+    char const* ptr_h_bytes = h_bytes.data();
+    out_sink_->host_write(ptr_h_bytes, total_num_bytes);
+  }
+}
+
+void writer::impl::write(table_view const& table,
+                         const table_metadata* metadata,
+                         cudaStream_t stream)
+{
+  CUDF_EXPECTS(table.num_columns() > 0 && table.num_rows() > 0, "Empty table.");
+
+  // write header: column names separated by delimiter:
+  //
+  write_chunked_begin(table, metadata, stream);
+
+  // no need to check same-size columns constraint; auto-enforced by table_view
+  auto n_rows_per_chunk = options_.rows_per_chunk();
+  //
+  // This outputs the CSV in row chunks to save memory.
+  // Maybe we can use the total_rows*count calculation and a memory threshold
+  // instead of an arbitrary chunk count.
+  // The entire CSV chunk must fit in CPU memory before writing it out.
+  //
+  if (n_rows_per_chunk % 8)  // must be divisible by 8
+    n_rows_per_chunk += 8 - (n_rows_per_chunk % 8);
+  CUDF_EXPECTS(n_rows_per_chunk > 0, "write_csv: invalid chunk_rows; must be at least 8");
+
+  auto exec = rmm::exec_policy(stream);
+
+  auto num_rows = table.num_rows();
+  std::vector<size_type> splits;
+
+  if (num_rows <= n_rows_per_chunk)
+    splits.push_back(num_rows);
+  else {
+    auto n_chunks = num_rows / n_rows_per_chunk;
+    splits.resize(n_chunks);
+
+    rmm::device_vector<size_type> d_splits(n_chunks, n_rows_per_chunk);
+    thrust::inclusive_scan(exec->on(stream), d_splits.begin(), d_splits.end(), d_splits.begin());
+
+    CUDA_TRY(cudaMemcpyAsync(splits.data(),
+                             d_splits.data().get(),
+                             n_chunks * sizeof(size_type),
+                             cudaMemcpyDeviceToHost,
+                             stream));
+
+    CUDA_TRY(cudaStreamSynchronize(stream));
+  }
+
+  // split table_view into chunks:
+  //
+  auto vector_views = cudf::experimental::split(table, splits);
+
+  // convert each chunk to CSV:
+  //
+  for (auto&& sub_view : vector_views) {
+    std::vector<std::unique_ptr<column>> str_column_vec;
+    column_to_strings_fn converter{options_, mr_};
+
+    // populate vector of string-converted columns:
+    //
+    std::transform(sub_view.begin(),
+                   sub_view.end(),
+                   std::back_inserter(str_column_vec),
+                   [converter](auto const& current_col) {
+                     return cudf::experimental::type_dispatcher(
+                       current_col.type(), converter, current_col);
+                   });
+
+    // create string table view from str_column_vec:
+    //
+    auto str_table_ptr = std::make_unique<cudf::experimental::table>(std::move(str_column_vec));
+    table_view str_table_view{std::move(*str_table_ptr)};
+
+    // concatenate columns in each row into one big string column
+    //(using null representation and delimiter):
```

Comment:
I think this is the right decision and should be much more efficient than removing nulls for each column beforehand.

Reply:
Agree.